feature/SOF-6597-1: support for POSCAR file import to MD

VsevolodX opened the pull request:
Import of multiple files, handling results from Made.parsers.convertFromNative(): alert of error, successful addition of material on correct JSON materialConfig

VsevolodX committed:
feat: import of multiple files

VsevolodX committed:
cancel button closes dialog

VsevolodX committed:
errors from parser shown in alert

VsevolodX committed:
temp check while conversion is not implemented yet

VsevolodX committed:
fetch dependency from wip branch

VsevolodX committed:
commit hash added to include intermediate changes

VsevolodX commented on src/components/header_menu/HeaderMenuToolbar.jsx:
```diff
@@ -288,9 +285,10 @@ class HeaderMenuToolbar extends React.Component {
                 onHide={() => this.setState({ showImportMaterialsDialog: false })}
                 onSubmit={(materials) => {
-                    onAdd(materials);
+                    if (materials) onAdd(materials);
```
Temporary measure. So that app doesn't freeze while feature is not implemented

VsevolodX committed:
code.js required as devdependency

VsevolodX committed:
version of code.js that is used by made.js

VsevolodX committed:
update to package-lock.json

VsevolodX committed:
fix: remove temporary workaround

VsevolodX committed:
update: package.json update to use package in dev

VsevolodX committed:
update: package-lock.json update to use commit version of made.js

Created by npm i of edited package.json. It's the only way to do that. Manual surgical edits to package-lock.json didn't give the desired result.

VsevolodX committed:
feat: drag&drop files upload with remove options

VsevolodX committed:
feat: add datagrid to display imported files

VsevolodX committed:
update: show format of imported files

VsevolodX committed:
update: improve ux

VsevolodX committed:
update: import of my commit of made.js

unsigned6 commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,323 @@
+import nativeFormats from "@exabyte-io/made.js/lib/parsers/native_formats";
+import Button from "@mui/material/Button";
+import Dialog from "@mui/material/Dialog";
+import DialogActions from "@mui/material/DialogActions";
+import DialogContent from "@mui/material/DialogContent";
+import DialogTitle from "@mui/material/DialogTitle";
```
To ommit code duplication we already created generic component with defined DialogActions, DialogContent and DialogTitle: https://github.com/Exabyte-io/exabyte-web-app/blob/dev/src/application/imports/core_cove/components/material-ui/Dialog.tsx. Please use it here as well (button labels are configurable, have a look on other this component usages, like this: https://github.com/Exabyte-io/exabyte-web-app/blob/dev/src/application/imports/science_meve/components/pseudopotential/PseudoUploadDialog.tsx).

VsevolodX replied:
Links lead to page not found for me. 
Also, can I adjust the height and width of the dialog? I reused ActionsDialog defined in standalone MD, but was unable to change styles of it from within my component.

VsevolodX replied:
This component looks great. Especially the ability to specify maxWidth explicitly. But the problem is it doesn't exist in the open-source version of material designer. Not sure if it's possible to import it.

unsigned6 commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,323 @@
+
+import { Material } from "../../material";
+
+class DefaultImportModalDialog extends React.Component {
```
we decided that preferable to write functional components every where, adjust this.

VsevolodX replied:
Ok. I prefer them too.

timurbazhirov replied:
It's OK to use this as a class - in MD we were using classes

VsevolodX replied:
I'll use as a class

unsigned6 commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,323 @@
+import { Material } from "../../material";
+
+class DefaultImportModalDialog extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            fileNames: [],
+            formats: [],
+            texts: [],
+            dragging: false,
+        };
```
let's use `useState` hook for each state field separate setter

VsevolodX replied:
Agree on this as well.

unsigned6 commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,323 @@
+            texts: [],
+            dragging: false,
+        };
+        this.title = "Import Materials";
```
should be passed as prop

unsigned6 commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,323 @@
+import nativeFormats from "@exabyte-io/made.js/lib/parsers/native_formats";
```
we already have TypeScript in this project, preferable to write new components using .ts/.tsx extensions.

timurbazhirov replied:
Where exactly is TS in Materials Designer? I don't think we are using it here yet

unsigned6 commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,323 @@
+            },
+        ];
+
+        const dropZoneStyle = {
+            height: "160px",
+            width: "560px",
+            display: "flex",
+            flexDirection: "column",
+            justifyContent: "center",
+            alignItems: "center",
+            textAlign: "center",
+            justifyText: "center",
+            border: "2px dashed gray",
+            borderRadius: "5px",
+            cursor: "pointer",
+            backgroundColor: dragging && "grey",
+        };
+        const paperStyle = {
+            position: "absolute",
+            top: "10%",
+            width: "800px",
+        };
+
+        const dataGridStyle = {
+            height: "400px",
+            backgroundColor: dragging && "grey",
+        };
+
```
This could be defined outside of the component

VsevolodX replied:
I'll move it outside. btw, is it ok to have styles defined as js object, or is it better to move them to .scss files?

timurbazhirov replied:
It's OK to keep here for now

unsigned6 commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,323 @@
+    show: PropTypes.bool.isRequired,
+    onClose: PropTypes.func.isRequired,
+    onSubmit: PropTypes.func.isRequired,
+    // material: PropTypes.object.isRequired,
```
we try to use in new components with TS just interfaces for props validation, adjust this please.

VsevolodX committed:
chore: clean up

k0stik commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,320 @@
+import nativeFormats from "@exabyte-io/made.js/lib/parsers/native_formats";
+import {
```
Let's split this into separate imports according to the MUI [Docs](https://mui.com/material-ui/guides/minimizing-bundle-size/#option-one-use-path-imports)
`import Button from "@mui/material/Button"`
`import Dialog from "@mui/material/Dialog"`
etc.

Unfortunately, we are not able to setup webpack plugin  in the web-app for now, and that leads to a huge js bundle in the production environment

VsevolodX replied:
Ok

k0stik commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,320 @@
+    DialogTitle,
+    FormControl,
+} from "@mui/material";
+import { DataGrid } from "@mui/x-data-grid";
```
Same here if possible

timurbazhirov replied:
+1

VsevolodX replied:
When I import as DataGrid then I get an Error:
'@mui/x-data-grid' does not contain a default export (imported as 'DataGrid').

VsevolodX committed:
feat: add tests

Initial commit to save progress. These tests fail currently.

VsevolodX committed:
update: clean up according to PR comments

VsevolodX committed:
update: changes to run first test

VsevolodX committed:
update: more tests that pass

VsevolodX committed:
feat: add file upload method to exabrowser

For some reason files paths concatenated with "\n" aren't uploaded. Error: file not found on your system

timurbazhirov commented on src/components/include/DefaultImportModalDialog.jsx:
```diff
@@ -0,0 +1,320 @@
+import nativeFormats from "@exabyte-io/made.js/lib/parsers/native_formats";
```
let's call this `nativeFormatParsers`

timurbazhirov replied:
And import Made directly

VsevolodX replied:
renamed to nativeFormatParsers (and in made.js native_format_parsers.js respectively), will import Made directly

VsevolodX committed:
feat: tests for add-remove-import of files

VsevolodX committed:
update: refactor and simplify files handling

VsevolodX committed:
update: remove unused import

VsevolodX committed:
update: rename mehod

VsevolodX committed:
update: remove unused import

VsevolodX committed:
update: add dependency to use a function from

VsevolodX committed:
chore: clean up

VsevolodX committed:
update: package.json and package-lock.json updates

I do use function from /code.js explicitly, so it needed to be installed

VsevolodX committed:
fix: address build test fail (wild guess)

VsevolodX committed:
Update: update regarding import name change

timurbazhirov commented on tests/cucumber/features/menu/input-output/add-remove-import-files.feature:
```diff
@@ -0,0 +1,50 @@
+    Then the DefaultImportModalDialog should be closed
+
+    # Reset
+    When I delete materials with index "2"
```
Newline, pls https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file

VsevolodX replied:
Other files did get the newline on save for some reason. Only this one didn't (.feature extension might be he cause?)

timurbazhirov commented on tests/cucumber/features/menu/edit/reset-clone-undo-redo.feature:
```diff
@@ -45,4 +45,4 @@ Feature: User can create a material, then undo the changes and/or reset to the i
       | path    | index   |
       | si-clone.json | $INT{2} |
-
+    When I undo
```
Remove

VsevolodX replied:
Oops. That's probably slipped during cmd+F search

timurbazhirov commented on tests/cucumber/features/menu/input-output/add-remove-import-files.feature:
```diff
@@ -0,0 +1,50 @@
+@watch
```
Remove

VsevolodX committed:
fix: add newline

VsevolodX committed:
update: add correct end lines

VsevolodX committed:
update: remove @watch

VsevolodX committed:
update: use new version of made.js
